Add lint rule for requiring useEffect deps

ahuth opened the pull request:
[sc-224374]

This PR adds a lint rule for requiring `useEffect` deps.

I'm 50/50 on whether we should introduce this or not, because
- There are valid reasons for not having a deps array (e.g. we want to run something after every render/update)
- But I think in our case it's often an oversight or copypasta (and if it was intentional we can safely suppress the lint rule)

Let me know your thoughts 🙏

ahuth committed:
Convert eslint-config-edu-react into a plugin

So we can add some custom rules to it.

ahuth committed:
Add use-effect-deps-presence lint rule

ahuth commented on packages/eslint-plugin-edu-react/package.json:
```diff
@@ -1,23 +1,23 @@
 {
-  "name": "@chanzuckerberg/eslint-config-edu-react",
+  "name": "@chanzuckerberg/eslint-plugin-edu-react",
```
To add a custom rule I had to convert this from a "config" to a "plugin" (eslint is kind of weird, and "configs" are just config, but "plugins" can have config and new rules).

An alternative would be to separate into both a config and a plugin, but that seems like too much work and not enough benefit to me.

ahuth commented on packages/eslint-plugin-edu-react/src/index.ts:
```diff
@@ -7,6 +13,7 @@ module.exports = {
     },
   },
   rules: {
+    '@chanzuckerberg/eslint-plugin-edu-react/use-effect-deps-presence': 'error',
```
Maybe this should be

```suggestion
    '@chanzuckerberg/eslint-plugin-edu-react/use-effect-deps-presence': 'warning',
```

🤔

ahuth committed:
Remove unnecessary config from tests

booc0mtaco commented:
I think I would be in favor of it being a warning to start. Assuming the main penalty is potential performance impacts, a warning could notify individuals about this, and permit those deps-array-less scenarios without an ignore comment. It's also a good code review topic (e.g., "can you explain why no deps array is needed here")

Are there any other worse outcomes I'm ignoring?

dierat commented:
>There are valid reasons for not having a deps array (e.g. we want to run something after every render/update)

I think that in the case something does need to run again on every update,
- that's rare enough that it's not that big of a deal to add a lint override comment
- we would probably want someone to leave a comment explaining it anyway

I'm fine with the lint rule, and I'm thinking of it as a reminder to consider how you actually want your code to run.

ahuth commented:
@booc0mtaco

> Are there any other worse outcomes I'm ignoring?

Not really. Worst case the logic is different than intended, but that's a danger for any code 😂 . Although it may be particularly easy to forget the deps array and get different logic.

booc0mtaco commented:
@ahuth ah gotcha. I bias conservative on introducing restrictions, so after some more thought, I change to keeping introducing it as error, as it seems not to be too bad to be more strict. We could always loosen it later if it does cause annoyance

ahuth committed:
Make use-effect-deps-presence a warning instead of error

timzchang commented:
I think some of this has already been said, so to add/retell a few thoughts:

Some potential tradeoffs of adding this rule:
✅ - protects against people adding useEffect without a dependency array unintentionally. In >95%(?) of cases, one would want there to be a dependency array.
❌ - it's not traditional to have an error for something that's not technically wrong, and people might be confused

I don't really think the con is something to be that concerned about since the error message is very clear and actionable. So overall feel good about this change!

ahuth committed:
Add eslint-plugin-edu-react as a plugin in the recommended config

palasha reviewed (commented):
What does this add that `react-hooks/exhaustive-deps` doesnt?

ahuth committed:
Tweak eslint-plugin-edu-react to actually work

Not sure why, but it seems for things to work in real apps we need
- To use module.exports
- To use the short name when defining plugins and rules

ahuth commented:
> What does this add that react-hooks/exhaustive-deps doesnt?

Given a deps array, `exhaustive-deps` makes sure it has all required entries in it.

But it doesn't ensure that there's a deps array in the first place.

ahuth committed:
Revert "Make use-effect-deps-presence a warning instead of error"

This reverts commit 98f736291f00cddfab6e3e09072537220e087eb0.

ahuth commented on packages/eslint-plugin-edu-react/src/index.ts:
```diff
@@ -1,13 +1,22 @@
+    // prettier-ignore
+    '@chanzuckerberg/edu-react/use-effect-deps-presence': 'error',
+    // TODO: Turn this rule on when it's released. It's not present in v6.6.1.
+    'jsx-a11y/anchor-ambiguous-text': 'off',
```
While testing this I discovered that this rule isn't actually released, yet 😅

ahuth commented on packages/eslint-plugin-edu-react/src/index.ts:
```diff
@@ -1,13 +1,22 @@
   rules: {
-    'jsx-a11y/anchor-ambiguous-text': 'error',
+    // prettier-ignore
+    '@chanzuckerberg/edu-react/use-effect-deps-presence': 'error',
```
Weirdly the "full" name (`eslint-plugin-edu-react` vs `edu-react`) doesn't work. No idea why (docs indicate it should).
